File: go/ancestor_relations_test.go
package apidef

import (
	"encoding/json"
	"os"
	"reflect"
	"testing"
)

func TestAncestorSource(t *testing.T) {
	data, err := os.ReadFile("../ts/test/ancestor-relations.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []struct {
		Entity, Model map[string]any
		Relations     string
	}
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatal(err)
	}
	for _, row := range cases {
		before, _ := json.Marshal(row.Entity)
		model, source := entityAncestorSource(row.Entity)
		if !reflect.DeepEqual(model, row.Model) || source != row.Relations {
			t.Fatalf("ancestor source: model=%#v source=%q", model, source)
		}
		after, _ := json.Marshal(row.Entity)
		if string(before) != string(after) {
			t.Fatal("builder mutated model")
		}
	}
}

func TestInferredAncestorTargets(t *testing.T) {
	data, err := os.ReadFile("../ts/test/ancestor-targets.json")
	if err != nil {
		t.Fatal(err)
	}
	var row struct{ Entities, Expected map[string]any }
	if err := json.Unmarshal(data, &row); err != nil {
		t.Fatal(err)
	}
	filterEntityAncestors(row.Entities)
	actual, _ := json.Marshal(row.Entities)
	expected, _ := json.Marshal(row.Expected)
	if string(actual) != string(expected) {
		t.Fatalf("targets: %s, want %s", actual, expected)
	}
}

File: ts/test/ancestor-relations.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import Path from 'node:path'
import { Aontu } from 'aontu'
import { entityAncestorSource } from '../dist/builder/entity/entity'
import { filterEntityAncestors } from '../dist/transform/entity'

const root = Path.resolve(__dirname, '../..')
const schema = readFileSync(Path.join(root, 'model/apidef.aon'), 'utf8')

test('inferred ancestors retain only existing other entities', () => {
  const row = JSON.parse(readFileSync(Path.join(root, 'ts/test/ancestor-targets.json'), 'utf8'))
  filterEntityAncestors(row.entities)
  assert.deepEqual(row.entities, row.expected)
})

test('ancestor source preserves chains and does not mutate the model', () => {
  const cases = JSON.parse(readFileSync(Path.join(root, 'ts/test/ancestor-relations.json'), 'utf8'))
  for (const row of cases) {
    const before = JSON.stringify(row.entity)
    assert.deepEqual(entityAncestorSource(row.entity), { model: row.model, relations: row.relations })
    assert.equal(JSON.stringify(row.entity), before)
  }
  const { relations } = entityAncestorSource(cases[0].entity)
  const model = new Aontu().generate(schema + '\nmain:kit:entity:{galaxy:{} planet:{} moon:{' + relations + '}}')
  assert.deepEqual(model.main.kit.entity.moon.relations.ancestors,
    [['$.main.kit.entity.planet'], ['$.main.kit.entity.galaxy', '$.main.kit.entity.planet']])
})

test('ancestors must link to existing other entities', () => {
  for (const target of ['missing', 'moon', 'planet.fields']) {
    const src = schema + '\nmain:kit:entity:{planet:{} moon:relations:ancestors:[[path($.main.kit.entity.' + target + ')]]}'
    assert.throws(() => new Aontu().generate(src), /aontu\/(rel_unresolved|constraint)/)
  }
  for (const value of ['path($.main.kit.flow.other)', '"planet"', '"$.main.kit.entity.planet"']) {
    const src = schema + '\nmain:kit:flow:other:{}\nmain:kit:entity:{planet:{} moon:relations:ancestors:[[' + value + ']]}'
    assert.throws(() => new Aontu().generate(src), /aontu\/(rel_address|scalar_value)/)
  }
})
